public/app/features/alerting/unified/components/receivers/TemplateForm.tsx

```diff
@@ -0,0 +1,153 @@
+
+interface Values {
+  name: string;
+  content: string;
+}
+
+const defaults: Values = Object.freeze({
+  name: '',
+  content: '',
+});
+
+interface Props {
+  existing?: Values;
+  config: AlertManagerCortexConfig;
+  alertManagerSourceName: string;
+}
+
+export const TemplateForm: FC<Props> = ({ existing, alertManagerSourceName, config }) => {
+  const styles = useStyles2(getStyles);
+  const dispatch = useDispatch();
+
+  useCleanup((state) => state.unifiedAlerting.saveAMConfig);
+
+  const { loading, error } = useUnifiedAlertingSelector((state) => state.saveAMConfig);
+
+  const submit = (values: Values) => {
+    // wrap content in "define" if it's not already wrapped, in case user did not do it/
+    // it's not obvious that this is needed for template to work
+    const content = ensureDefine(values.name, values.content);
+
+    // add new template to template map
+    const template_files = {
+      ...config.template_files,
+      [values.name]: content,
+    };
+
+    // delete existing one (if name changed, otherwise it was overwritten in previous step)
+    if (existing && existing.name !== values.name) {
+      delete template_files[existing.name];
+    }
+
+    // make sure name for the template is configured on the alertmanager config object
+    const templates = [
+      ...(config.alertmanager_config.templates ?? []).filter((name) => name !== existing?.name),
+      values.name,
+    ];
+
+    const newConfig: AlertManagerCortexConfig = {
+      template_files,
+      alertmanager_config: {
+        ...config.alertmanager_config,
+        templates,
+      },
+    };
+    dispatch(updateAlertManagerConfigAction({ alertManagerSourceName, newConfig, oldConfig: config }));
+  };
+
+  const { handleSubmit, register, errors } = useForm<Values>({
+    mode: 'onSubmit',
+    defaultValues: existing ?? defaults,
+  });
+
+  const validateNameIsUnique: Validate = (name: string) => {
+    return !config.template_files[name] || existing?.name === name
+      ? true
+      : 'Another template with this name already exists.';
+  };
+
+  return (
+    <form onSubmit={handleSubmit(submit)}>
+      <h4>{existing ? 'Edit message template' : 'Create message template'}</h4>
+      {error && (
+        <Alert severity="error" title="Error saving template">
+          {error.message || (error as any)?.data?.message || String(error)}
+        </Alert>
+      )}
+      <Field label="Template name" error={errors?.name?.message} invalid={!!errors.name?.message}>
+        <Input
+          width={42}
+          autoFocus={true}
+          ref={register({
+            required: { value: true, message: 'Required.' },
+            validate: { nameIsUnique: validateNameIsUnique },
+          })}
+          name="name"
+        />
+      </Field>
+      <Field
+        description={
+          <>
+            You can use the{' '}
+            <a href="https://pkg.go.dev/text/template?utm_source=godoc" target="__blank">
```

Comment:
good to add rel="noreferrer" for external links

Reply:
done!